### ISSUE TITLE
[health 7.0.1] Incorrect behaviour when working with permissions

### ISSUE THREAD

**Issue by muezz** [cph-cachet/flutter-plugins#781]
## Emulator and OS
 - Device: iPhone 14 Plus and iPhone 12 Plus
 - OS: iOS 16.2

# The bug and steps to reproduce
### Scenarios:
1. Fresh device which is running the app for the first time (after erasing the data and settings or creating a new emulator device).
2. After accepting all required permissions (through `requestAuthorization` or manually in the settings).
3. After the user has given the permissions, he closes the app, goes to the settings, removes all or some of the permissions and relaunches the app.
## Problem with `hasPermissions`
If I run the following piece of code after any of the three scenarios above, I always get `null`:
```dart
var types = [
    HealthDataType.ACTIVE_ENERGY_BURNED,
    HealthDataType.STEPS,
    HealthDataType.WEIGHT,
    HealthDataType.EXERCISE_TIME,
    HealthDataType.WORKOUT,
  ];
  var permissions = [
    HealthDataAccess.READ_WRITE,
    HealthDataAccess.READ_WRITE,
    HealthDataAccess.READ_WRITE,
    HealthDataAccess.READ,
    HealthDataAccess.READ_WRITE
  ];
  final hasPermissions = await health.hasPermissions(
    types,
    permissions: permissions,
  );
  log(hasPermissions.toString());
```
This means that there is no way for me to know if the user has these permissions or not.
## Problem with `requestAuthorization`
In a similar way, if I try to `requestAuthorization`, I always get true regardless of the scenario. I don't understand why there is no differentiation here as well.
## Problem with `HealthDataType.EXERCISE_TIME`
In the code snippet above, you can see that I am trying to request only `READ` permission for `EXERCISE_TIME`. I have noticed that if I try to access `READ_WRITE` or `WRITE`, the app crashes completely without any exception or error.

# Expected behavior
In the three scenarios, following should be the expected behavior in my point of view (unless there is some explicit shortcoming from iOS).
1. `hasPermissions` should be `false` and `requestAuthorization` should return `true` if the user accepts all permissions on prompt and `false` otherwise.
2. `hasPermissions` should be `true` and `requestAuthorization` should return `true`.
3. `hasPermissions` should be `false` and `requestAuthorization` should return `false` as the user cannot be prompted again and must be asked to accept permissions manually. Here the developer should take the liberty to `openAppSettings()`.

### Flutter doctor
```console 
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.10.5, on macOS 13.4.1 22F82 darwin-arm64, locale en-PK)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] VS Code (version 1.80.1)
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
```

**Comment by hoffmatteo** [comment]
Similarly for requestAuthorization:
Apple does not provide whether the request was actually approved, since it does not share READ permissions, see here: https://developer.apple.com/documentation/healthkit/hkhealthstore/1614152-requestauthorization.
The only thing that can be returned, and that is returned, is whether the request was successfully made, which is what the true returns.

**Comment by muezz** [comment]
@hoffmatteo In that case, what is the recommendation here? How can I ever know if the data I am fetching is correct or not.

Perhaps I can forget about it because if there is permission, I will get the data and if not, the returned array will be empty. In that case, it would be on the user to realize that they have to give permissions. When I say this, I mean the scenarios where the user may provide all permissions initially and then remove some or all by going to the settings.

This is okay-ish but I do not like that as it is a very bad UX experience. I would like to have the option to tell the user that they need to add all permissions in the settings if some are missing (which I would like to check using `hasPermissions`).

I realize that this may be out of our hands since the Apple gods have deemed it so. In any case, what is the recommended approach?

**Comment by hoffmatteo** [comment]
I agree on iOS this is a tricky subject, this is why our health plugin simply reflects the system behaviour. In the end it does depend on the user not changing permissions, since there is no option to tell if they have added all READ permissions, so it has to be solved in a UI way. How exactly that looks depends more on your own application, maybe someone else can chime in with their use case.

For HealthDataType.EXERCISE_TIME I cannot replicate your behaviour, if you execute the code in the example app does this still persist?

**Comment by hoffmatteo** [comment]
Closed as stale.

**Comment by Muzammil-Bit** [comment]
One of the best-formatted issues and responses. Thanks a lot for these!

How could we solve this issue with UI? Does anyone has any use case or suggestion?
